### ISSUE TITLE
Questions about reproducing stage 2

### ISSUE THREAD

**Issue by huangcanan** [VITA-MLLM/LUCY#10]
First of all, thank you very much for the open-source work. I am currently working on reproducing the model based on this work, skipping stage 1 and directly proceeding with stage 2 training. I downloaded the open-source audio encoder [here](https://huggingface.co/VITA-MLLM/LUCY-Audio-Encoder-110kh) as the pre-trained model, and then directly conducted s2p0 training using the open-source [AudioQA-1M](https://huggingface.co/datasets/VITA-MLLM/AudioQA-1M) as the training data, which I reconstructed it into 24k wav files using [snac.](https://github.com/hubertsiuzdak/snac) . I have three questions:

1. Approximately how many steps are needed for the s2p0 training to converge?
2. To what level will the converged loss decrease?
3. After the s2p0 converges, can the model successfully conduct text response and voice response?

During my training of s2p0, I have completed 60k+ steps. The loss value has decreased to around 5.5, showing a trend of convergence. However, the text responses still have many errors, and it is unable to synthesize normal voice. Some of my key training parameters are as follows:

TASKS="AQACONVA RQACONVA RQACONV"

nnodes=1
nproc_per_node=8
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 

vita/scripts/train.py \
--deepspeed config/zero2.json\
--model_type "qwen2" \
--initialize_additional_modules True \
--model_name_or_path $MODEL_NAME_OR_PATH \
--audio_encoder $AUDIO_ENCODER \
--audio_projector_type "identity" \
--freeze_backbone True \
--freeze_audio_encoder_adapter True \
--freeze_audio_encoder True \
--freeze_tts_adapter False \
--freeze_embed_tokens False \
--per_device_train_batch_size 32 \
--per_device_eval_batch_size 32 \
--add_codec_target True \
--num_train_epochs 5 \
--load_best_model_at_end True \
--save_steps 1000 \
--save_total_limit 20 \
--eval_strategy "steps" \
--eval_steps 1000 \
--learning_rate 2e-5 \
--weight_decay 0. \
--warmup_ratio 0.03 \
--logging_steps 10 \
--lr_scheduler_type "cosine" \
--gradient_checkpointing True \
--bf16 True \
--model_hidden_size 1536 \
--audio_encoder_hidden_size 1024 \
--audio_projector_hidden_size 7168 \
--audio_num_codebook 7 \
--text_vocab_size 152064 \
--text_special_tokens 64 \
--audio_vocab_size 4096 \
--audio_special_tokens 64 \
--data_jsons $DATA_JSONS \
--eval_data_jsons $EVAL_DATA_JSONS \
--text_additional "EOT" "PAD_T" "BOT" "ANS_T" "TTS" "TQA" "TQAA" \
--audio_additional "EOA" "PAD_A" "BOA" "ANS_A" "ASR" "AQA" "AQAA" "M29" "F10" "ER" \
--asr_template /mnt/data/hetinggao/manifest/asr_prompts/asr_template.json \
--tasks ${TASKS} \
--output_dir ${OUTPUT_DIR} \
--sample_rate 24000 \
--audio_feature_rate 50 \
--dataloader_num_workers 8 \
--remove_unused_columns False \
--max_keep_sample_size $((25*16000)) \
--tune_text_embed True \
--tie_word_embeddings True \
--loss_reduction mean \
--max_input_length 1500 \
--use_last_turn_if_codec True \

**Comment by BBQLXDM** [comment]
> First of all, thank you very much for the open-source work. I am currently working on reproducing the model based on this work, skipping stage 1 and directly proceeding with stage 2 training. I downloaded the open-source audio encoder [here](https://huggingface.co/VITA-MLLM/LUCY-Audio-Encoder-110kh) as the pre-trained model, and then directly conducted s2p0 training using the open-source [AudioQA-1M](https://huggingface.co/datasets/VITA-MLLM/AudioQA-1M) as the training data, which I reconstructed it into 24k wav files using [snac.](https://github.com/hubertsiuzdak/snac) . I have three questions:
> 
> 1. Approximately how many steps are needed for the s2p0 training to converge?
> 2. To what level will the converged loss decrease?
> 3. After the s2p0 converges, can the model successfully conduct text response and voice response?
> 
> During my training of s2p0, I have completed 60k+ steps. The loss value has decreased to around 5.5, showing a trend of convergence. However, the text responses still have many errors, and it is unable to synthesize normal voice. Some of my key training parameters are as follows:
> 
> TASKS="AQACONVA RQACONVA RQACONV"
> 
> nnodes=1 nproc_per_node=8 CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
> 
> vita/scripts/train.py --deepspeed config/zero2.json --model_type "qwen2" --initialize_additional_modules True --model_name_or_path $MODEL_NAME_OR_PATH --audio_encoder $AUDIO_ENCODER --audio_projector_type "identity" --freeze_backbone True --freeze_audio_encoder_adapter True --freeze_audio_encoder True --freeze_tts_adapter False --freeze_embed_tokens False --per_device_train_batch_size 32 --per_device_eval_batch_size 32 --add_codec_target True --num_train_epochs 5 --load_best_model_at_end True --save_steps 1000 --save_total_limit 20 --eval_strategy "steps" --eval_steps 1000 --learning_rate 2e-5 --weight_decay 0. --warmup_ratio 0.03 --logging_steps 10 --lr_scheduler_type "cosine" --gradient_checkpointing True --bf16 True --model_hidden_size 1536 --audio_encoder_hidden_size 1024 --audio_projector_hidden_size 7168 --audio_num_codebook 7 --text_vocab_size 152064 --text_special_tokens 64 --audio_vocab_size 4096 --audio_special_tokens 64 --data_jsons $DATA_JSONS --eval_data_jsons $EVAL_DATA_JSONS --text_additional "EOT" "PAD_T" "BOT" "ANS_T" "TTS" "TQA" "TQAA" --audio_additional "EOA" "PAD_A" "BOA" "ANS_A" "ASR" "AQA" "AQAA" "M29" "F10" "ER" --asr_template /mnt/data/hetinggao/manifest/asr_prompts/asr_template.json --tasks ${TASKS} --output_dir ${OUTPUT_DIR} --sample_rate 24000 --audio_feature_rate 50 --dataloader_num_workers 8 --remove_unused_columns False --max_keep_sample_size $((25*16000)) --tune_text_embed True --tie_word_embeddings True --loss_reduction mean --max_input_length 1500 --use_last_turn_if_codec True \

Hello, I'm also trying to reproduce Lucy. Currently, I'm working on reproducing stage 2. I'd like to ask how you converted the snac file into a wav file. Could you please provide the code?

**Comment by huangcanan** [comment]
> > First of all, thank you very much for the open-source work. I am currently working on reproducing the model based on this work, skipping stage 1 and directly proceeding with stage 2 training. I downloaded the open-source audio encoder [here](https://huggingface.co/VITA-MLLM/LUCY-Audio-Encoder-110kh) as the pre-trained model, and then directly conducted s2p0 training using the open-source [AudioQA-1M](https://huggingface.co/datasets/VITA-MLLM/AudioQA-1M) as the training data, which I reconstructed it into 24k wav files using [snac.](https://github.com/hubertsiuzdak/snac) . I have three questions:
> > 
> > 1. Approximately how many steps are needed for the s2p0 training to converge?
> > 2. To what level will the converged loss decrease?
> > 3. After the s2p0 converges, can the model successfully conduct text response and voice response?
> > 
> > During my training of s2p0, I have completed 60k+ steps. The loss value has decreased to around 5.5, showing a trend of convergence. However, the text responses still have many errors, and it is unable to synthesize normal voice. Some of my key training parameters are as follows:
> > TASKS="AQACONVA RQACONVA RQACONV"
> > nnodes=1 nproc_per_node=8 CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
> > vita/scripts/train.py --deepspeed config/zero2.json --model_type "qwen2" --initialize_additional_modules True --model_name_or_path $MODEL_NAME_OR_PATH --audio_encoder $AUDIO_ENCODER --audio_projector_type "identity" --freeze_backbone True --freeze_audio_encoder_adapter True --freeze_audio_encoder True --freeze_tts_adapter False --freeze_embed_tokens False --per_device_train_batch_size 32 --per_device_eval_batch_size 32 --add_codec_target True --num_train_epochs 5 --load_best_model_at_end True --save_steps 1000 --save_total_limit 20 --eval_strategy "steps" --eval_steps 1000 --learning_rate 2e-5 --weight_decay 0. --warmup_ratio 0.03 --logging_steps 10 --lr_scheduler_type "cosine" --gradient_checkpointing True --bf16 True --model_hidden_size 1536 --audio_encoder_hidden_size 1024 --audio_projector_hidden_size 7168 --audio_num_codebook 7 --text_vocab_size 152064 --text_special_tokens 64 --audio_vocab_size 4096 --audio_special_tokens 64 --data_jsons $DATA_JSONS --eval_data_jsons 

> Hello, I'm also trying to reproduce Lucy. Currently, I'm working on reproducing stage 2. I'd like to ask how you converted the snac file into a wav file. Could you please provide the code?

use this [script](https://github.com/VITA-MLLM/LUCY/blob/main/vita/util/snac_decode.py)

**Comment by Hertin** [comment]
s2p0 updates only the embeddings and it is normal the loss is high. 1 epoch of training should be enough. The loss will futher decrease in s2p1 when you start to update the llm. It tooks about 20k step in s2p1 for the loss to decrease to below 4 and the model will be able to produce normal audio after that.

**Comment by BBQLXDM** [comment]
> > > First of all, thank you very much for the open-source work. I am currently working on reproducing the model based on this work, skipping stage 1 and directly proceeding with stage 2 training. I downloaded the open-source audio encoder [here](https://huggingface.co/VITA-MLLM/LUCY-Audio-Encoder-110kh) as the pre-trained model, and then directly conducted s2p0 training using the open-source [AudioQA-1M](https://huggingface.co/datasets/VITA-MLLM/AudioQA-1M) as the training data, which I reconstructed it into 24k wav files using [snac.](https://github.com/hubertsiuzdak/snac) . I have three questions:
> > > 
> > > 1. Approximately how many steps are needed for the s2p0 training to converge?
> > > 2. To what level will the converged loss decrease?
> > > 3. After the s2p0 converges, can the model successfully conduct text response and voice response?
> > > 
> > > During my training of s2p0, I have completed 60k+ steps. The loss value has decreased to around 5.5, showing a trend of convergence. However, the text responses still have many errors, and it is unable to synthesize normal voice. Some of my key training parameters are as follows:
> > > TASKS="AQACONVA RQACONVA RQACONV"
> > > nnodes=1 nproc_per_node=8 CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
> > > vita/scripts/train.py --deepspeed config/zero2.json --model_type "qwen2" --initialize_additional_modules True --model_name_or_path $MODEL_NAME_OR_PATH --audio_encoder $AUDIO_ENCODER --audio_projector_type "identity" --freeze_backbone True --freeze_audio_encoder_adapter True --freeze_audio_encoder True --freeze_tts_adapter False --freeze_embed_tokens False --per_device_train_batch_size 32 --per_device_eval_batch_size 32 --add_codec_target True --num_train_epochs 5 --load_best_model_at_end True --save_steps 1000 --save_total_limit 20 --eval_strategy "steps" --eval_steps 1000 --learning_rate 2e-5 --weight_decay 0. --warmup_ratio 0.03 --logging_steps 10 --lr_scheduler_type "cosine" --gradient_checkpointing True --bf16 True --model_hidden_size 1536 --audio_encoder_hidden_size 1024 --audio_projector_hidden_size 7168 --audio_num_codebook 7 --text_vocab_size 152064 --text_special_tokens 64 --audio_vocab_size 4096 --audio_special_tokens 64 --data_jsons $DATA_JSONS --eval_data_jsons
> 
> > Hello, I'm also trying to reproduce Lucy. Currently, I'm working on reproducing stage 2. I'd like to ask how you converted the snac file into a wav file. Could you please provide the code?
> 
> use this [script](https://github.com/VITA-MLLM/LUCY/blob/main/vita/util/snac_decode.py)

thanks
